pkg/operator/ceph/object/shared_pools.go

```diff
@@ -0,0 +1,612 @@
+package object
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+const (
+	defaultPlacementName           = "default"
+	defaultPlacementCephConfigName = "default-placement"
+	defaultPlacementStorageClass   = "STANDARD"
+)
+
+func IsNeedToCreateObjectStorePools(sharedPools cephv1.ObjectSharedPoolsSpec) bool {
+	for _, pp := range sharedPools.PoolPlacements {
+		if pp.Name == defaultPlacementName {
+			// No need to create pools. External pools from default placement will be used
+			return false
+		}
+	}
+	if sharedPools.MetadataPoolName != "" && sharedPools.DataPoolName != "" {
+		// No need to create pools. Shared pools will be used
+		return false
+	}
+	return true
+}
+
+func validatePoolPlacements(placements []cephv1.PoolPlacementSpec) error {
+	names := make(map[string]struct{}, len(placements))
+	for _, p := range placements {
+		if _, ok := names[p.Name]; ok {
+			return fmt.Errorf("invalidObjStorePoolCofig: invalid placement %s: placement names must be unique", p.Name)
+		}
+		names[p.Name] = struct{}{}
+		if err := validatePoolPlacementStorageClasses(p.StorageClasses); err != nil {
+			return fmt.Errorf("invalidObjStorePoolCofig: invalid placement %s: %w", p.Name, err)
+		}
+	}
+	return nil
+}
+
+func validatePoolPlacementStorageClasses(scList []cephv1.PlacementStorageClassSpec) error {
+	names := make(map[string]struct{}, len(scList))
+	for _, sc := range scList {
+		if sc.Name == defaultPlacementStorageClass {
+			return fmt.Errorf("invalid placement StorageClass %q: %q name is reserved", sc.Name, defaultPlacementStorageClass)
+		}
+		if _, ok := names[sc.Name]; ok {
+			return fmt.Errorf("invalid placement StorageClass %q: name must be unique", sc.Name)
+		}
+		names[sc.Name] = struct{}{}
+	}
+	return nil
+}
+
+func adjustZonePlacementPools(zone map[string]interface{}, spec cephv1.ObjectSharedPoolsSpec) (hasChanged bool, err error) {
```

Comment:
I would prefer to not modify the `zone` param in-place, as I have had the experience that this can lead to bugs that are hard to track down. 

However, I think I've realized why you haven't done this. IIUC, it's challenging to return a modified copy of `zone` instead of modifying zone in-place. My memory is that there is no generic deepcopy function for a `map[string]interface{}`, and not great language tools to do it manually. 

If it's feasible to do this without in-place modification, I'd like to see that. If not, would you make a comment at the top of the function about why it's modified in place for future reference?

Reply:
We can safely make a deepcopy by marshal/unmarshal zone to json. I think that it will be appropriate solution here because performance doesn't matter too much hare.

Reply:
now it makes deep copy with json marshl/unmarshl and returns updated obj